src/ipv4/static_ipv4.ml

```diff
@@ -64,90 +64,72 @@ module Make (R: Mirage_random.C) (C: Mirage_clock.MCLOCK) (Ethernet: Mirage_prot
-      let hdr =
-        let src = match src with None -> t.ip | Some x -> x in
-        let off = if fragment then 0x0000 else 0x4000 in
-        Ipv4_packet.{
-          options = Cstruct.empty ;
-          src ; dst ;
-          ttl ; off ; id = 0 ;
-          proto = Ipv4_packet.Marshal.protocol_to_int proto }
-      in
-      let writeout size fill =
-        Ethernet.write t.ethif mac `IPv4 ~size fill >|= function
-        | Error e ->
-          Log.warn (fun f -> f "Error sending Ethernet frame: %a" Ethernet.pp_error e);
-          Error (`Ethif e)
-        | Ok () -> Ok ()
-      in
-      Log.debug (fun m -> m "ip write: mtu is %d, hdr_len is %d, size %d payload len %d, needed_bytes %d"
-                   mtu hdr_len size (Cstruct.lenv bufs) needed_bytes) ;
-      if mtu >= needed_bytes then begin
-        (* single fragment *)
+      if not fragment && multiple then
+        Lwt.return (Error `Would_fragment)
+      else
+        let off =
+          match fragment, multiple with
+          | true, true -> 0x2000
+          | false, false -> 0x4000
+          | true, false -> 0x0000
+          | false, true -> assert false (* handled by conditional above *)
+        in
+        let hdr =
+          let src = match src with None -> t.ip | Some x -> x in
+          let id = if multiple then Randomconv.int16 R.generate else 0 in
+          Ipv4_packet.{
+            options = Cstruct.empty ;
+            src ; dst ; ttl ; off ; id ;
+            proto = Ipv4_packet.Marshal.protocol_to_int proto }
+        in
+        let writeout size fill =
+          Ethernet.write t.ethif mac `IPv4 ~size fill >|= function
+          | Error e ->
+            Log.warn (fun f -> f "Error sending Ethernet frame: %a"
+                         Ethernet.pp_error e);
+            Error (`Ethif e)
+          | Ok () -> Ok ()
+        in
+        Log.debug (fun m -> m "ip write: mtu is %d, hdr_len is %d, size %d \
+                               payload len %d, needed_bytes %d"
+                      mtu hdr_len size (Cstruct.lenv bufs) needed_bytes) ;
+        let leftover = ref Cstruct.empty in
+        (* first fragment *)
         let fill buf =
-          let hdr_buf, payload_buf = Cstruct.split buf hdr_len in
+          let payload_buf = Cstruct.shift buf hdr_len in
           let header_len = headerf payload_buf in
           if header_len > size then begin
             Log.err (fun m -> m "headers returned length exceeding size") ;
             invalid_arg "headerf exceeds size"
           end ;
           (* need to copy the given payload *)
-          let len, leftover =
+          let len, rest =
             Cstruct.fillv ~src:bufs ~dst:(Cstruct.shift payload_buf header_len)
           in
-          if leftover <> [] then begin
-            Log.err (fun m -> m "there's some leftover data") ;
-            invalid_arg "leftover data"
-          end ;
+          leftover := Cstruct.concat rest;
```

Comment:
Hmm this seems a bit expensive, would it be problematic to have `leftover` be a list of `Cstruct.t` to avoid having to allocate this for each fragment?

Reply:
I suspect this is covered by the commit message. yes, a performance improvement would be to redefine `Fragments.fragment` to copy less. no, I don't see a specific reason to do this (unless I see numbers showing that this is actually a bottleneck -- as mentioned earlier, most packets are not fragmented.)